File: windows/tests/SSMConnect.Workflow.Tests/Fixtures/PortRecorder.cs
using System.Text.Json;
using SSMConnect.Domain;

namespace SSMConnect.Workflow.Tests.Fixtures;

/// <summary>One recorded call to an injected port.</summary>
public sealed record RecordedCall(string Port, string Method, IReadOnlyDictionary<string, object?> Arguments)
{
    public override string ToString() => $"{Port}.{Method}";
}

/// <summary>
/// Records every port call and serves the outcome queue a fixture declared for it.
/// </summary>
/// <remarks>
/// Queue semantics are part of the contract: each entry is consumed by one call, and the last entry
/// repeats. That is what lets a fixture say "fails once, then succeeds" without any scripting.
/// </remarks>
public sealed class PortRecorder(Given given)
{
    private readonly Dictionary<string, int> _consumed = [];
    private readonly List<RecordedCall> _calls = [];
    private readonly Lock _gate = new();

    public IReadOnlyList<RecordedCall> Calls
    {
        get
        {
            lock (_gate)
            {
                return [.. _calls];
            }
        }
    }

    public int CountOf(string port, string method) =>
        Calls.Count(call => call.Port == port && call.Method == method);

    /// <summary>Records a call and returns the outcome the fixture declared, throwing if it is an error.</summary>
    public JsonElement? Invoke(string port, string method, params (string Name, object? Value)[] arguments)
    {
        lock (_gate)
        {
            _calls.Add(new RecordedCall(port, method, arguments.ToDictionary(a => a.Name, a => a.Value)));
        }

        Outcome? outcome = NextOutcome(port, method);
        if (outcome is null)
        {
            return null;
        }

        if (outcome.Error is OutcomeError error)
        {
            throw ToException(error);
        }

        return outcome.Result;
    }

    private Outcome? NextOutcome(string port, string method)
    {
        if (!given.Ports.TryGetValue(port, out Dictionary<string, Outcome[]>? methods) ||
            !methods.TryGetValue(method, out Outcome[]? queue) ||
            queue.Length == 0)
        {
            return null;
        }

        string key = $"{port}.{method}";
        lock (_gate)
        {
            _consumed.TryGetValue(key, out int index);
            _consumed[key] = index + 1;
            return queue[Math.Min(index, queue.Length - 1)];
        }
    }

    /// <summary>
    /// Maps a portable error kind onto this implementation's exception type. The fixture never names
    /// a platform type, so this mapping is the whole translation layer.
    /// </summary>
    private static Exception ToException(OutcomeError error) => error.Kind switch
    {
        "expiredCredentials" => new ExpiredCredentialsException(),
        "signInRequired" => new SignInRequiredException(),
        "invalidRegion" => new ProfileConfigurationException("resource region", error.Message ?? string.Empty),
        "instanceTerminated" => new InstanceTerminatedException(error.Message ?? "i-unknown"),
        "stageTimeout" => new StageTimeoutException(error.Message ?? "Stage"),
        "localPortInUse" => new TunnelException(error.Message ?? "The local port is already in use."),
        "tunnelNotEstablished" => DcvReadinessException.TunnelNotEstablished(0),
        "dcvServerNotReady" => DcvReadinessException.DcvServerNotReady(0),
        "agentUnauthorized" => new AgentException(error.Message ?? "The agent rejected the request.", responded: true),
        "agentUnreachable" => new AgentException(error.Message ?? "The agent is unreachable.", responded: false),
        "viewerNotInstalled" => new InvalidOperationException("Amazon DCV Viewer is not installed."),
        "awsServiceError" => new AwsServiceException(error.Message ?? "AWS returned an error."),
        _ => new InvalidOperationException(error.Message ?? "Unknown failure."),
    };
}

File: SSMConnectKit/Tests/SSMConnectKitTests/Conformance/PortRecorder.swift
import Foundation
@testable import SSMConnectDomain
@testable import SSMConnectWorkflow
@testable import SSMConnectAWS
@testable import SSMConnectMacOS
@testable import SSMConnectUI

/// One recorded call to an injected port.
struct RecordedCall: CustomStringConvertible {
    let port: String
    let method: String
    let arguments: [String: JSONValue]

    var description: String { "\(port).\(method)" }
}

/// Records every port call and serves the outcome queue the fixture declared for it.
///
/// The queue semantics are part of the contract, not an implementation detail: each entry is
/// consumed by one call and the last entry repeats, which is how a fixture says "fails once, then
/// succeeds" with no scripting.
final class PortRecorder: @unchecked Sendable {
    private let ports: [String: [String: [Outcome]]]
    private let lock = NSLock()
    private var consumed: [String: Int] = [:]
    private var recorded: [RecordedCall] = []

    init(given: Given) { self.ports = given.ports }

    var calls: [RecordedCall] {
        lock.lock(); defer { lock.unlock() }
        return recorded
    }

    func count(port: String, method: String) -> Int {
        calls.filter { $0.port == port && $0.method == method }.count
    }

    /// Record a call and return the declared result, throwing if the fixture declared an error.
    @discardableResult
    func invoke(_ port: String, _ method: String, _ arguments: [String: JSONValue] = [:]) throws -> JSONValue? {
        lock.lock()
        recorded.append(RecordedCall(port: port, method: method, arguments: arguments))
        let key = "\(port).\(method)"
        let index = consumed[key] ?? 0
        consumed[key] = index + 1
        lock.unlock()

        guard let queue = ports[port]?[method], !queue.isEmpty else { return nil }
        let outcome = queue[min(index, queue.count - 1)]
        if let error = outcome.error { throw Self.error(from: error) }
        return outcome.result
    }

    /// Maps a portable error kind onto the Swift error the flow actually catches. The fixture never
    /// names a platform type, so this mapping is the entire translation layer — and getting it
    /// wrong here silently changes which code path a case exercises.
    static func error(from declared: OutcomeError) -> Error {
        switch declared.kind {
        case "expiredCredentials":
            // Deliberately a real unmodeled AWS service error carrying `ExpiredTokenException`,
            // so the *shipping* `defaultExpiredCredentialsCheck` predicate is what recognises it.
            // A bespoke sentinel would have proved only that the harness agrees with itself.
            return unmodeledAWSServiceError(
                typeName: "ExpiredTokenException",
                message: declared.message ?? "The security token included in the request is expired",
                statusCode: .forbidden
            )
        case "signInRequired":
            return AuthError.signInRequired
        case "invalidRegion":
            return ProfileConfigError.invalidRegion(field: "resource region", value: declared.message ?? "")
        case "instanceTerminated":
            return EC2Error.instanceTerminated(instanceId: declared.message ?? "i-unknown")
        case "stageTimeout":
            return StageTimeoutError(stage: declared.message ?? "Stage")
        case "localPortInUse":
            return TunnelError.localPortInUse(port: 8443, pid: nil, processName: nil)
        case "tunnelNotEstablished":
            return DCVReadinessError.tunnelNotEstablished(port: 8443)
        case "dcvServerNotReady":
            return DCVReadinessError.dcvServerNotReady(port: 8443)
        case "agentUnauthorized":
            return AgentError.unauthorized
        case "agentUnreachable":
            // A transport failure, NOT an `AgentError`: the flow retries this while the freshly
            // opened agent tunnel settles, and treats a real agent response as final.
            return URLError(.cannotConnectToHost)
        case "viewerNotInstalled":
            return DCVError.viewerNotInstalled
        case "awsServiceError":
            return unmodeledAWSServiceError(
                typeName: "ServiceUnavailableException",
                message: declared.message ?? "AWS returned an error.",
                statusCode: .serviceUnavailable
            )
        default:
            return DCVError.launchFailed(reason: declared.message ?? "Unknown failure.")
        }
    }
}
